REPL-docview: fallback to doc(typeof(x)) for objects x

mgkurtz opened the pull request:
Fix #45154, e.g. for `x=1` let `?x` give the doc string of `Int`.

As I got somewhat confused by the level of indirection, I also added a comment to provide an overview over the internal control flow of the help mode implementation.

mgkurtz committed:
REPL-docview: fallback to doc(typeof(x)) for objects x

Fix #45154.

mgkurtz committed:
Fix typo in function name

mortenpi reviewed (commented):
How do the REPL outputs compare exactly? The help output is changed for both variables without docstrings and literals?

mgkurtz commented on stdlib/REPL/src/docview.jl:
```diff
@@ -176,10 +181,18 @@ function doc(binding::Binding, sig::Type = Union{})
+        if alias != binding
+            return doc(alias, sig)
+        elseif defined(binding) && !(resolve(binding) isa Union{Type, Module, Function})
+            println("`", binding, "` is of type `", typeof(resolve(binding)), "`\n")
```
Yes. Putting it as a normal paragraph also felt wrong. So, I tend to either get rid of this line entirely or prepend it in some emphasized flavor (bold?) to the markdown result.

mortenpi replied:
I think some explanation is necessary here for the user. If it just falls back to the docstring of the type, it looks like a bug. I think I'd just try prepending it as a paragraph + horizontal rule to the returned Markdown and see how that looks.

mgkurtz replied:
I have implemented your suggestion now. The horizontal rule gets added implicitly by the `insert_hlines` method.

mgkurtz commented:
> How do the REPL outputs compare exactly? The help output is changed for both variables without docstrings and literals?

Literals are handled in [`lookup_doc`](https://github.com/mgkurtz/julia/blob/7b9bfc072208059043d1e9af6ad715945994bd0f/stdlib/REPL/src/docview.jl#L230) which I did not change. We could of course add the “This is of type …” line there as well or remove it in both cases. (See my comment above.)

mgkurtz committed:
Move `lookup_doc` function to where it logically belongs

mgkurtz committed:
Change `summary`; incorporate heading in `MD` in `doc`

`summary` now returns `nothing` instead of recursing, if given
something, that is neither Module, nor Type or Function.

For help on objects, prepend a heading like "`a` is of type `Int64`" to
the returned `Markdown.MD`.

mgkurtz committed:
Fix previous commit

mgkurtz commented:
Sorry for leaving this stale for a while. I am not totally confident with all my changes. So please feel free, to criticize ☺️
